encoding/varint.ts

```diff
@@ -1,9 +1,166 @@
+ * `offset` defaults to `0`.
+ *
+ * If passed `buf` then that will be written into, starting at `offset`. The
+ * resulting returned `Uint8Array` will be a slice of `buf`.
+ *
+ * ```ts
+ * import { encodeU32 } from "https://deno.land/std@$STD_VERSION/encoding/varint.ts";
+ *
+ * encodeU32(1) == Uint8Array.of(0x7F, 0x02);
+ * encodeU32(300) == Uint8Array.of(0x7F, 0x02);
+ * ```
+ */
+export function encodeU32(
+  num: number,
+  buf: Uint8Array = new Uint8Array(MaxVarIntLen64),
+  offset = 0,
+): Uint8Array {
+  if (!Number.isInteger(num)) throw new TypeError("Floats are not supported");
+  if (num < 0) throw new RangeError("Signed integers are not supported");
+  if (num > U32MAX) {
+    throw new RangeError(
+      `The given number exceeds the limit of unsigned integer: ${num}`,
+    );
+  }
+  for (
+    let i = offset, len = Math.min(buf.length, MaxVarIntLen64);
+    i <= len;
+    i += 1
+  ) {
+    if (num < MSBN) {
+      buf[i] = Number(num);
+      i += 1;
+      return buf.slice(offset, i);
+    }
+    buf[i] = Number((num & 0xFF) | MSB);
+    num >>= SHIFT;
+  }
+  return buf;
+}
+
+/**
+ * Takes unsigned number `num` and converts it into a VarInt encoded
+ * `Uint8Array`, returning a `Uint8Array` slice of the
+ * encoded VarInt.
+ *
+ * If `buf` is not given then a Uint8Array will be created.
+ * `offset` defaults to `0`.
+ *
+ * If passed `buf` then that will be written into, starting at `offset`. The
+ * resulting returned `Uint8Array` will be a slice of `buf`
+ *
+ * ```ts
+ * import { encodeU64 } from "https://deno.land/std@$STD_VERSION/encoding/varint.ts";
+ *
+ * encodeU64(4294967295n) === Uint8Array.of(0xFF, 0xFF, 0xFF, 0xFF, 0x0F);
+ * ```
+ */
+export function encodeU64(
+  num: bigint,
+  buf: Uint8Array = new Uint8Array(MaxVarIntLen64),
+  offset = 0,
+): Uint8Array {
+  if (num < 0) throw new RangeError("Signed integers are not supported");
+  if (num > U64MAX) {
+    throw new RangeError(
+      `The given number exceeds the limit of unsigned long integer: ${num}`,
+    );
+  }
+  for (
+    let i = offset, len = Math.min(buf.length, MaxVarIntLen64);
+    i <= len;
+    i += 1
+  ) {
+    if (num < MSBN) {
+      buf[i] = Number(num);
+      i += 1;
+      return buf.slice(offset, i);
+    }
+    buf[i] = Number((num & 0xFFn) | MSBN);
+    num >>= SHIFTN;
+  }
+  return buf;
+}
 
 /**
- * Functions for encoding typed integers in array buffers.
+ * Given a `buf`, starting at `offset` (default: 0), begin decoding bytes as
+ * VarInt encoded bytes, for a maximum of 5 bytes (offset + 5). The return
+ * is of the decoded varint 32-bit number.
  *
- * @module
+ * VarInts are _not 32-bit by default_ so this should only be used in cases
+ * where the varint is _assured_ to be 32-bits.
+ *
+ * To know how many bytes the VarInt took to encode, simply negate `offset`
+ * from the returned new `offset`.
  */
+export function decodeU32(buf: Uint8Array, offset = 0): number {
```

Comment:
You’ll definitely want to somehow return how many bytes the varint was encoded in, so that consumers know which byte offset to continue reading data from.

Also, thanks ☺️

Reply:
Good point. In that case, we'll have to deprecate the current API as the form of the return values will change. @kt3k, are you in favour?

Reply:
I agree that we should return the used bytes somehow.

> we'll have to deprecate the current API as the form of the return values will change

I wonder if there's other design possibilities other than returning 2-tuple (Creating an array for each decode sounds a little expensive to me). For example, `npm:varint` sets the used bytes at `varint.decode.bytes` ref. https://github.com/chrisdickinson/varint#varintdecodebytes

How about exporting used bytes like `export let decodedBytes` and setting that value at each decode?

Reply:
My vote would be against magic variables or properties on functions as they can release Zalgo in async contexts (though I don't have a compelling usecase for unpacking varints async :joy:)

One option could be to simply add additional methods like `decodeU32AndGetOffset(): [number, number]`. I'm sure you can think of a better name though :smiley:

Reply:
I doubt that could cause zalgo as `decode` doesn't involve any async operations. I think the code like the below works reliably:

```js
const n = decode(...);
const bytes = SOME_MAGIC_VAR;
```

> One option could be to simply add additional methods like decodeU32AndGetOffset(): [number, number]. I'm sure you can think of a better name though 😃

This seems obvious and a good option too 👍

Reply:
I was thinking more about if you had something like:

```js
const n = decode(...);
await someOp
const bytes = SOME_MAGIC_VAR
```

I can't think of a reason why you'd need to await between extraction... perhaps if you're streaming responses... but then you can shuffle the lines around without much difficulty. It's just a bit of an awkward api.

Reply:
> I agree that we should return the used bytes somehow.
> 
> > we'll have to deprecate the current API as the form of the return values will change
> 
> I wonder if there's other design possibilities other than returning 2-tuple (Creating an array for each decode sounds a little expensive to me). For example, `npm:varint` sets the used bytes at `varint.decode.bytes` ref. https://github.com/chrisdickinson/varint#varintdecodebytes
> 
> How about exporting used bytes like `export let decodedBytes` and setting that value at each decode?

How is it expensive to return a 2-tuple? I don't understand how it is.

If it turns out to be inexpensive to return a 2-tuple, would it be better to return `{ result: number, offset: number }` for `decodeX()` and `{ buffer: UInt8Array, offset: number }` for `encodeX()` instead?

We should introduce these as `encode32Bit()`, `decode64Bit()`, etc., in conjunction with the deprecation.

Reply:
@keithamus 

> I can't think of a reason why you'd need to await between extraction... perhaps if you're streaming responses... but then you can shuffle the lines around without much difficulty. It's just a bit of an awkward api.

Ok. Now I agree that magic var design is confusing

Reply:
@iuioiua 

> We should introduce these as encode32Bit(), decode64Bit(), etc., in conjunction with the deprecation.

If we change the API names, how about following `x/varint` naming style? (`decode`, `decode32`, `encode`). Fortunately they don't have overlaps with the existing APIs, and also they look simpler than the current APIs.

Reply:
Ok, we'll do a straight port of x/varint. However, I think it's worth splitting `encode()` into `encode32()` and `encode64()` if there are performance gains to be had. I'll create a new issue for the exploration of this idea.